cocos/base/CCRef.h

```diff
@@ -123,6 +123,51 @@ class CC_DLL Ref
      * @js NA
      */
     unsigned int getReferenceCount() const;
+    
+    /**
+     * Returns a shared_ptr of this Ref.
+     *
+     * @returns A shared_ptr of this as a Ref.
+     * @js NA
+     * @lua NA
+     */
+    std::shared_ptr<cocos2d::Ref> shared_ptr() {
+        if(!_self_weak_ref.expired()) {
+            return _self_weak_ref.lock();
+        }
+        
+        auto shared_ptr = std::shared_ptr<cocos2d::Ref>(this, [=](cocos2d::Ref* ref)
+        {
+            if(ref->getReferenceCount() == 0) {
+                delete ref;
+            }
+        });
+        _self_weak_ref = shared_ptr;
+        return shared_ptr;
+    }
+    
+    /**
+     * Template version of shared_ptr
+     *
+     * @returns A shared_ptr of this type template.
+     * @js NA
+     * @lua NA
+     */
+    template <class T> std::shared_ptr<T> shared_ptr() {
+        if(!_self_weak_ref.expired()) {
+            return std::static_pointer_cast<T>(_self_weak_ref.lock());
+        }
+        
+        auto shared_ptr = std::shared_ptr<cocos2d::Ref>(this, [=](cocos2d::Ref* ref)
```

Comment:
Do we need capture on this lambda? Can't it just be [] (cocos2d::Ref* ref) ...

Reply:
Yes, it's superfluous.  I just did for readability.